File: pcm2pwle/beating_detection.py
#  Copyright 2026 Google LLC
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

"""Utility functions for detecting beating signals."""

import logging

import numpy as np
from scipy import fft
from scipy import signal

from pcm2pwle import common_utils
from pcm2pwle import control_point_extraction

sosfiltfilt = signal.sosfiltfilt
hilbert = signal.hilbert
find_peaks = signal.find_peaks
butter = signal.butter
rfft = fft.rfft
rfftfreq = fft.rfftfreq


def is_beating(
    data: np.ndarray,
    rate: int,
    peak_ratio_threshold: float = 2,
    freq_diff_threshold: float = 60,
    peak_energy_ratio_threshold: float = 0.1,
    energy_outside_peaks_ratio_threshold: float = 0.1,
) -> None | tuple[float, float]:
  """Checks if the given data is a beating signal.

  Args:
    data: The signal data to be checked.
    rate: The sampling rate of the data.
    peak_ratio_threshold: The threshold for the ratio of the FFT amplitudes at
      the two most prominent component frequencies.  The threshold for the ratio
      of the amplitude of the two peaks.
    freq_diff_threshold: The threshold for the frequency difference between the
      two most prominent component frequencies.
    peak_energy_ratio_threshold: The threshold for the ratio of the spectral
      energy at the two most prominent component frequencies to the total
      spectral energy.
    energy_outside_peaks_ratio_threshold: The threshold for the ratio of
      spectral energy outside the two most prominent component frequencies to
      the total spectral energy.

  Returns:
    None if the data is not a beating signal, otherwise returns a tuple of the
    two beat frequencies.
  """
  if data.size == 0:
    return None

  #### 1. FREQUENCY DOMAIN ANALYSIS ####
  xf = rfftfreq(len(data), 1 / rate)
  yf = rfft(data)
  yf = np.abs(yf) / np.max(np.abs(yf))

  sos = butter(5, rate / 10 / (rate / 2), btype='low', output='sos')
  yf = sosfiltfilt(sos, yf)

  peaks, properties = find_peaks(yf, prominence=0)

  # Not enough prominent peaks to create a beat
  if len(peaks) < 2:
    logging.warning(
        'Found %d peaks, not enough prominent peaks to create a beat',
        len(peaks),
    )
    return None

  prominences = properties['prominences']
  sorted_peak_indices = np.argsort(prominences)[::-1]
  peak1_idx, peak2_idx = peaks[sorted_peak_indices[:2]]

  freq1, freq2 = xf[peak1_idx], xf[peak2_idx]
  logging.info('freq1: %.2f Hz, freq2: %.2f Hz', freq1, freq2)
  fft_beat_freq = abs(freq1 - freq2)

  # Check if the frequency difference is within the plausible range for a beat
  logging.info('Frequency difference (%.2f Hz)', fft_beat_freq)
  if not (1 <= fft_beat_freq <= freq_diff_threshold):
    logging.warning(
        'Frequency difference (%.2f Hz) out of range [1 Hz, %d Hz]',
        fft_beat_freq,
        freq_diff_threshold,
    )
    return None

  # Check if the amplitude of the two peaks are close
  peak1_amp, peak2_amp = yf[peak1_idx], yf[peak2_idx]
  peak_ratio = peak1_amp / peak2_amp
  logging.info('peak1_amp: %.2f, peak2_amp: %.2f', peak1_amp, peak2_amp)
  logging.info('peak ratio: %.2f', peak_ratio)
  if peak_ratio > peak_ratio_threshold:
    logging.warning('peak ratio (%.2f) too large', peak_ratio)
    return None

  # Check peak energy is larger than a ratio of the spectrum energy
  peak_energy = yf[peak1_idx] ** 2 + yf[peak2_idx] ** 2
  spectrum_energy = np.sum(yf**2)
  peak_energy_ratio = peak_energy / spectrum_energy
  logging.info('peak energy ratio (%.2f)', peak_energy_ratio)
  if peak_energy_ratio < peak_energy_ratio_threshold:
    logging.warning('peak energy ratio (%.2f) too small', peak_energy_ratio)
    return None

  # Check if no energy outside of the two peaks
  spectrum_energy = np.sum(yf**2)
  freq_left = min(freq1, freq2) - 50
  freq_right = max(freq1, freq2) + 50
  energy_outside_peaks = np.sum(yf[(xf < freq_left) | (xf > freq_right)] ** 2)
  energy_outside_peaks_ratio = energy_outside_peaks / spectrum_energy
  logging.info('energy outside peaks ratio (%.2f)', energy_outside_peaks_ratio)
  if energy_outside_peaks_ratio > energy_outside_peaks_ratio_threshold:
    logging.warning(
        'energy outside peaks ratio (%.2f) too large',
        energy_outside_peaks_ratio,
    )
    return None

  #### 2. AMPLITUDE ENVELOPE ANALYSIS ####
  ## Check amplitude envelope valley point's freq are outliers
  analytic_signal = hilbert(data)
  amp_envelope = np.abs(analytic_signal)
  ins_phase = np.unwrap(np.angle(analytic_signal))
  freq_envelope = np.diff(ins_phase) / (2 * np.pi) * rate
  freq_envelope = np.insert(freq_envelope, 0, freq_envelope[0])

  chunks = common_utils.zero_out_low_amplitude_chunks(
      data,
      rate,
      amplitude_threshold_fraction=0.01,
      duration_threshold_ms=20,
  )
  for _, (start, end) in enumerate(chunks):
    amp_envelope[start:end] = 0.0
    freq_envelope[start:end] = 0.0

  logging.info(
      'envelope frequency: %.2f Hz, repeat distance: %.2f',
      fft_beat_freq,
      1 / fft_beat_freq * rate * 0.95,
  )
  valley_idxs, _ = find_peaks(
      -amp_envelope,
      distance=1 / fft_beat_freq * rate * 0.95,
      prominence=np.max(np.abs(amp_envelope)) * 0.1,
  )
  valley_freq = freq_envelope[valley_idxs]
  mean_valley_freq = np.mean(np.abs(valley_freq))
  freq_q3 = np.quantile(np.abs(freq_envelope), 0.75)
  freq_q1 = np.quantile(np.abs(freq_envelope), 0.25)
  freq_iqr = freq_q3 - freq_q1
  outlier_freq = freq_q3 + 1.5 * freq_iqr
  logging.info(
      'mean valley frequency: %.2f Hz, outlier frequency: %.2f Hz',
      mean_valley_freq,
      outlier_freq,
  )
  avg_freq = (freq1 + freq2) / 2
  outlier_freq_low = avg_freq - 10
  outlier_freq_high = min(avg_freq * 2, outlier_freq)
  logging.info(
      'mean valley frequency: %.2f Hz, outlier frequency: %.2f Hz %.2f Hz',
      mean_valley_freq,
      outlier_freq_low,
      outlier_freq_high,
  )
  if (
      outlier_freq_low < mean_valley_freq < outlier_freq_high
      or min(abs(valley_freq)) == 0
  ):
    logging.warning(
        'mean valley frequency (%.2f Hz) is not outlier',
        mean_valley_freq,
    )
    return None

  return (freq1, freq2)


def extract_beating_control_points(
    amp_envelope: np.ndarray,
    freq_envelope: np.ndarray,
    beating_freqs: tuple[float, float],
    rate: int,
    error_threshold: float,
    jnd_multiple: float,
) -> list[common_utils.ControlPoint]:
  """Extracts the control points from the amplitude envelope and frequency envelope.

  Args:
    amp_envelope: Amplitude envelope of the signal.
    freq_envelope: Frequency envelope of the signal.
    beating_freqs: The two component frequencies of the beating signal.
    rate: Sampling rate of the signal.
    error_threshold: The error threshold for Ramer-Douglas-Peucker algorithm.
    jnd_multiple: The JND multiple for the control points.

  Returns:
    A list of ControlPoint.
  """
  carrier_freq = (beating_freqs[0] + beating_freqs[1]) / 2
  beating_freq = np.abs(beating_freqs[0] - beating_freqs[1])

  amp_fitted_downsampled, freq_fitted_downsampled = (
      common_utils.preprocess_envelopes(amp_envelope, freq_envelope, rate)
  )

  all_amp_points = control_point_extraction.rdp(
      amp_fitted_downsampled,
      epsilon=error_threshold,
      point_type=common_utils.PointType.AMPLITUDE,
  )

  # JND value default to 1.5dB * multiplier for vibration amplitude and
  # frequency.
  amp_points = control_point_extraction.rdp_with_min_distance(
      all_amp_points,
      min_duration_ms=common_utils.MIN_DURATION,
      perceiption_jnd=1.5 * jnd_multiple,
  )

  control_points = []
  for point in amp_points:
    control_points.append(
        common_utils.ControlPoint(
            time=point.time,
            amplitude=point.amplitude,
            frequency=carrier_freq,
            time_shift=point.time_shift,
        )
    )

  control_points = common_utils.zero_amplitude_at_zero_frequency(
      control_points, freq_fitted_downsampled
  )
  control_points = common_utils.set_fist_last_point_amp_to_zero(control_points)
  logging.info('Number of control points: %d', len(control_points))

  common_utils.reject_invalid_waveform(
      control_points, time_shift_threshold=100, mean_freq_threshold=25
  )
  common_utils.reject_invalid_beating_waveform(
      beating_freq, common_utils.MIN_DURATION
  )

  return control_points

File: pcm2pwle/beating_detection.cc
#include "pcm2pwle/beating_detection.h"

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <limits>
#include <optional>
#include <string_view>
#include <utility>
#include <vector>

#include "pcm2pwle/common_utils.h"
#include "pcm2pwle/control_point_extraction.h"
#include "pcm2pwle/dsp_utils.h"

namespace pcm2pwle {

static std::vector<int> LocalMaxima1d(const std::vector<double>& x) {
  std::vector<int> midpoints;
  int n = x.size();
  for (size_t i = 1; i + 1 < x.size(); ++i) {
    if (x[i - 1] < x[i]) {
      int iAhead = i + 1;
      while (iAhead < n - 1 && x[iAhead] == x[i]) {
        iAhead++;
      }
      if (x[iAhead] < x[i]) {
        int left = i;
        int right = iAhead - 1;
        midpoints.push_back((left + right) / 2);
        i = iAhead;
      }
    }
  }
  return midpoints;
}

static std::vector<int> SelectByPeakDistance(
    const std::vector<int>& peaks, const std::vector<double>& priority,
    double distance) {
  if (peaks.empty()) return {};
  std::vector<int> order(peaks.size());
  for (size_t i = 0; i < peaks.size(); ++i) order[i] = i;
  std::stable_sort(order.begin(), order.end(),
                   [&](int a, int b) { return priority[a] > priority[b]; });

  std::vector<bool> keep(peaks.size(), true);
  for (size_t i = 0; i < order.size(); ++i) {
    int idx = order[i];
    if (!keep[idx]) continue;
    int peakPos = peaks[idx];
    for (size_t j = 0; j < peaks.size(); ++j) {
      if (keep[j] && j != static_cast<size_t>(idx) &&
          std::abs(peaks[j] - peakPos) < distance) {
        keep[j] = false;
      }
    }
  }

  std::vector<int> result;
  for (size_t i = 0; i < peaks.size(); ++i) {
    if (keep[i]) {
      result.push_back(peaks[i]);
    }
  }
  return result;
}

static std::pair<std::vector<int>, std::vector<double>> ComputeProminences(
    const std::vector<double>& x, const std::vector<int>& peaks,
    double minProminence) {
  std::vector<int> kept;
  std::vector<double> proms;
  for (int p : peaks) {
    int lb = p;
    while (lb > 0 && x[lb - 1] <= x[p]) {
      lb--;
    }
    double lmin = x[p];
    for (int j = lb; j <= p; ++j) {
      if (x[j] < lmin) lmin = x[j];
    }

    int rb = p;
    while (rb + 1 < x.size() && x[rb + 1] <= x[p]) {
      rb++;
    }
    double rmin = x[p];
    for (int j = p; j <= rb; ++j) {
      if (x[j] < rmin) rmin = x[j];
    }

    double prom = x[p] - std::max(lmin, rmin);
    if (prom >= minProminence) {
      kept.push_back(p);
      proms.push_back(prom);
    }
  }
  return {kept, proms};
}

static std::vector<int> FindPeaks(const std::vector<double>& x,
                                  std::optional<double> distance,
                                  std::optional<double> prominence) {
  auto peaks = LocalMaxima1d(x);
  if (distance.has_value() && distance.value() >= 1.0) {
    std::vector<double> priority(peaks.size());
    for (size_t i = 0; i < peaks.size(); ++i) {
      priority[i] = x[peaks[i]];
    }
    peaks = SelectByPeakDistance(peaks, priority, distance.value());
  }
  if (prominence.has_value()) {
    auto [kept, proms] = ComputeProminences(x, peaks, prominence.value());
    peaks = kept;
  }
  return peaks;
}

std::optional<std::pair<double, double>> IsBeating(
    const std::vector<double>& data, int rate, double peakRatioThreshold,
    double freqDiffThreshold, double peakEnergyRatioThreshold,
    double energyOutsidePeaksRatioThreshold) {
  int n = data.size();
  if (n == 0) return std::nullopt;

  auto fftOut = Rfft(data);
  int numFreqs = fftOut.size();
  std::vector<double> yf(numFreqs);
  std::vector<double> xf(numFreqs);
  double maxYf = 0.0;
  for (int i = 0; i < numFreqs; ++i) {
    yf[i] = std::abs(fftOut[i]);
    if (yf[i] > maxYf) maxYf = yf[i];
    xf[i] = (static_cast<double>(i) * rate) / static_cast<double>(n);
  }

  if (maxYf > 0.0) {
    for (int i = 0; i < numFreqs; ++i) {
      yf[i] /= maxYf;
    }
  }

  auto sos = ButterworthLowpassSos(5, 0.2, 2.0);
  yf = Sosfiltfilt(sos, yf);

  auto initialPeaks = LocalMaxima1d(yf);
  auto [peaks, prominences] = ComputeProminences(yf, initialPeaks, 0.0);

  if (peaks.size() < 2) {
    return std::nullopt;
  }

  std::vector<int> sortedOrder(peaks.size());
  for (size_t i = 0; i < peaks.size(); ++i) sortedOrder[i] = i;
  std::sort(sortedOrder.begin(), sortedOrder.end(),
            [&](int a, int b) { return prominences[a] > prominences[b]; });

  int peak1Idx = peaks[sortedOrder[0]];
  int peak2Idx = peaks[sortedOrder[1]];

  double freq1 = xf[peak1Idx];
  double freq2 = xf[peak2Idx];
  double fftBeatFreq = std::abs(freq1 - freq2);

  if (fftBeatFreq < 1.0 || fftBeatFreq > freqDiffThreshold) {
    return std::nullopt;
  }

  double peak1Amp = yf[peak1Idx];
  double peak2Amp = yf[peak2Idx];
  double peakRatio = peak1Amp / (peak2Amp != 0.0 ? peak2Amp : 1e-6);
  if (peakRatio > peakRatioThreshold) {
    return std::nullopt;
  }

  double spectrumEnergy = 0.0;
  for (int i = 0; i < numFreqs; ++i) {
    spectrumEnergy += yf[i] * yf[i];
  }
  if (spectrumEnergy == 0.0) return std::nullopt;

  double peakEnergy = peak1Amp * peak1Amp + peak2Amp * peak2Amp;
  if (peakEnergy / spectrumEnergy < peakEnergyRatioThreshold) {
    return std::nullopt;
  }

  double freqLeft = std::min(freq1, freq2) - 50.0;
  double freqRight = std::max(freq1, freq2) + 50.0;
  double energyOutsidePeaks = 0.0;
  for (int i = 0; i < numFreqs; ++i) {
    if (xf[i] < freqLeft || xf[i] > freqRight) {
      energyOutsidePeaks += yf[i] * yf[i];
    }
  }
  if (energyOutsidePeaks / spectrumEnergy > energyOutsidePeaksRatioThreshold) {
    return std::nullopt;
  }

  std::vector<double> real, imag;
  Hilbert(data, &real, &imag);
  std::vector<double> ampEnvelope(n);
  std::vector<double> insPhase(n);
  for (size_t i = 0; i < data.size(); ++i) {
    ampEnvelope[i] = std::sqrt(real[i] * real[i] + imag[i] * imag[i]);
    insPhase[i] = std::atan2(imag[i], real[i]);
  }
  std::vector<double> unwrapped = Unwrap(insPhase);
  std::vector<double> freqEnvelope(n);
  for (size_t i = 1; i < data.size(); ++i) {
    freqEnvelope[i] = ((unwrapped[i] - unwrapped[i - 1]) / (2.0 * M_PI)) * rate;
  }
  freqEnvelope[0] = n > 1 ? freqEnvelope[1] : 0.0;

  auto chunks = ZeroOutLowAmplitudeChunks(data, rate, 0.01, 20.0);
  for (const auto& chunk : chunks) {
    for (int i = chunk.first; i < chunk.second && i < data.size(); ++i) {
      ampEnvelope[i] = 0.0;
      freqEnvelope[i] = 0.0;
    }
  }

  double minDistance = (1.0 / fftBeatFreq) * rate * 0.95;
  double maxAmp = 0.0;
  for (size_t i = 0; i < data.size(); ++i) {
    if (std::abs(ampEnvelope[i]) > maxAmp) maxAmp = std::abs(ampEnvelope[i]);
  }
  double valleyProminence = maxAmp * 0.1;

  std::vector<double> negAmp(n);
  for (size_t i = 0; i < data.size(); ++i) negAmp[i] = -ampEnvelope[i];

  auto valleyIdxs = FindPeaks(negAmp, minDistance, valleyProminence);
  if (valleyIdxs.empty()) return std::nullopt;

  double valleyFreqSum = 0.0;
  double minAbsValleyFreq = std::numeric_limits<double>::infinity();
  for (int idx : valleyIdxs) {
    double absF = std::abs(freqEnvelope[idx]);
    valleyFreqSum += absF;
    if (absF < minAbsValleyFreq) minAbsValleyFreq = absF;
  }
  if (minAbsValleyFreq == 0.0) return std::nullopt;
  double meanValleyFreq = valleyFreqSum / valleyIdxs.size();

  std::vector<double> sortedAbsFreq(n);
  for (size_t i = 0; i < data.size(); ++i)
    sortedAbsFreq[i] = std::abs(freqEnvelope[i]);
  std::sort(sortedAbsFreq.begin(), sortedAbsFreq.end());
  double freqQ1 = sortedAbsFreq[static_cast<size_t>(n * 0.25)];
  double freqQ3 = sortedAbsFreq[static_cast<size_t>(n * 0.75)];
  double freqIqr = freqQ3 - freqQ1;
  double outlierFreq = freqQ3 + 1.5 * freqIqr;
  double avgFreq = (freq1 + freq2) / 2.0;
  double outlierFreqLow = avgFreq - 10.0;
  double outlierFreqHigh = std::min(avgFreq * 2.0, outlierFreq);

  if (meanValleyFreq > outlierFreqLow && meanValleyFreq < outlierFreqHigh) {
    return std::nullopt;
  }

  return std::make_pair(freq1, freq2);
}

std::vector<ControlPoint> ExtractBeatingControlPoints(
    const std::vector<double>& ampEnvelope,
    const std::vector<double>& freqEnvelope,
    const std::pair<double, double>& beatingFreqs, int rate,
    std::string_view pwleType, double errorThreshold) {
  double carrierFreq = (beatingFreqs.first + beatingFreqs.second) / 2.0;
  double beatingFreq = std::abs(beatingFreqs.first - beatingFreqs.second);

  auto [ampFittedDownsampled, freqFittedDownsampled] =
      PreprocessEnvelopes(ampEnvelope, freqEnvelope, rate);

  auto allAmpPoints =
      Rdp(ampFittedDownsampled, errorThreshold, PointType::AMPLITUDE);
  auto ampPoints = RdpWithMinDistance(allAmpPoints, pwleType);

  std::vector<ControlPoint> controlPoints;
  controlPoints.reserve(ampPoints.size());
  for (const auto& point : ampPoints) {
    ControlPoint cp;
    cp.time = point.time;
    cp.amplitude = point.amplitude;
    cp.frequency = carrierFreq;
    cp.timeShift = point.timeShift;
    controlPoints.push_back(cp);
  }

  controlPoints =
      ZeroAmplitudeAtZeroFrequency(controlPoints, freqFittedDownsampled);
  controlPoints = SetFirstLastPointAmpToZero(controlPoints);

  RejectInvalidWaveform(controlPoints, 100.0, 25.0);
  RejectInvalidBeatingWaveform(beatingFreq);

  return controlPoints;
}

}  // namespace pcm2pwle
